qiskit/utils/optionals.py

```diff
@@ -0,0 +1,293 @@
+
+.. list-table::
+    :widths: 25 75
+
+    * - .. py:data:: HAS_CPLEX
+      - The `IBM CPLEX Optimizer <https://www.ibm.com/uk-en/analytics/cplex-optimizer>`__ is a
+        high-performance mathematical programming solver for linear, mixed-integer and quadratic
+        programming.  It is required by the :class:`.BIPMapping` transpiler pass.
+
+    * - .. py:data:: HAS_CVXPY
+      - `CVXPY <https://www.cvxpy.org/>`__ is a Python package for solving convex optimization
+        problems.  It is required for calculating diamond norms with
+        :func:`.quantum_info.diamond_norm`.
+
+    * - .. py:data:: HAS_DOCPLEX
+      - `IBM Decision Optimization CPLEX Modelling
+        <http://ibmdecisionoptimization.github.io/docplex-doc/>`__ is a library for prescriptive
+        analysis.  Like CPLEX, it is required for the :class:`.BIPMapping` transpiler pass.
+
+    * - .. py:data:: HAS_FIXTURES
+      - The test suite has additional features that are available if the optional `fixtures
+        <https://launchpad.net/python-fixtures>`__ module is installed.  This generally also needs
+        :data:`HAS_TESTTOOLS` as well.  This is generally only needed for Qiskit developers.
+
+    * - .. py:data:: HAS_IPYTHON
+      - If `the IPython kernel <https://ipython.org/>`__ is available, certain additional
+        visualisations and line magics are made available.
+
+    * - .. py:data:: HAS_IPYWIDGETS
+      - Monitoring widgets for jobs running on external backends can be provided if `ipywidgets
+        <https://ipywidgets.readthedocs.io/en/latest/>`__ is available.
+
+    * - .. py:data:: HAS_JAX
+      - Some methods of gradient calculation within :mod:`.opflow.gradients` require `JAX
+        <https://github.com/google/jax>`__ for autodifferentiation.
+
+    * - .. py:data:: HAS_MATPLOTLIB
+      - Qiskit Terra provides several visualisation tools in the :mod:`.visualization` module.
+        Almost all of these are built using `Matplotlib <https://matplotlib.org/>`__, which must
+        be installed in order to use them.
+
+    * - .. py:data:: HAS_NETWORKX
+      - Internally, Qiskit uses the high-performance `retworkx
+        <https://github.com/Qiskit/retworkx>`__ library as a core dependency, but sometimes it can
+        be convenient to convert things into the Python-only `NetworkX <https://networkx.org/>`__
+        format.  There are converter methods on :class:`.DAGCircuit` if NetworkX is present.
+
+    * - .. py:data:: HAS_NLOPT
+      - `NLOpt <https://nlopt.readthedocs.io/en/latest/>`__ is a nonlinear optimisation library,
+        used by the global optimizers in :mod:`.algorithms.optimizers`.  See installation details in
+        :ref:`installing-nlopt`.
+
+    * - .. py:data:: HAS_PIL
+      - PIL is a Python image-manipulation library.  Qiskit actually uses the `pillow
+        <https://pillow.readthedocs.io/en/stable/>`__ fork of PIL if it is available when generating
+        circuit visualisations in some modes.
```

Comment:
Besides the latex drawer, it's also used by the dag visualizations and anywhere we call to pydot or graphviz to wrap the generated image file in an object we can return.

Reply:
Thanks - I rewrote it a little bit to make it clearer that it's used in a few places.

3dbbe56ee